src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -1482,6 +1517,20 @@ DataTypePtr makeNullableRecursively(DataTypePtr type)
         return nested_type ? std::make_shared<DataTypeArray>(nested_type) : nullptr;
     }
 
+    if (which.isVariant())
+    {
+        const auto * variant_type = assert_cast<const DataTypeVariant *>(type.get());
+        DataTypes nested_types;
+        for (const auto & nested_type: variant_type->getVariants())
+        {
+            /// unlike tuple or array, here we do not want to make any of the variants nullable
+            /// so we do not call makeNullableRecursively
```

Comment:
We don't want to make variants Nullable, but we want to make nested types inside variants Nullable. For example, `Array(UInt64)` -> `Array(Nullable(UInt64))`. So, we should call `makeNullableRecursively` when `!nested_type->lowCardinality() && nested_type->haveSubtypes()`